### DISCUSSION TITLE
Static typing without `class_name` on every single custom scene

### DISCUSSION THREAD

**Discussion by SKFrozenCloud** [godotengine/godot-proposals#7279] [Scripting]
I want to enforce and use static typing without the need to put `class_name` on every single custom scene.

For example:
by putting `class_name CameraModule` in my custom camera scene I can then put in my main scene script:
`@onready var camera_module := $CameraModule as CameraModule`
...which enables me to have autocompleted functions and enforce the static typings from the camera scene.

But that means that even a rock needs a custom class if I want the static typing and autocompletetion:
`class_name RockEntity`
`@onready var rock_entity := $RockEntity as RockEntity`

In other words, **every single custom scene I create will need a `class_name`!** Even a small rock!

### New Proposal

Be able to write it as:
`var rock := $Rock as preload("rock.gd")`

**Comment by AThousandShips** [comment]
How do you propose having a named type without `class_name`?

Also what's the problem with having a single line naming your class?

You don't have to use named classes, but static typing can't be enforced easily on paths like this because you don't know at compile time what type it is, how could you?

**Reply by SKFrozenCloud** [reply]
The problem is that ever single custom scene such as a rock, hat, shadow, radio, clouds and so on need a custom class name. Not to mention if you have several types of clouds or rocks you it's even a bigger mess.

My proposal is simple:

Be able to specify what what scene the path is with the res: path.

In other words if I have a rock scene located at `res://rock.tscn`.
I can later write `@onready var rock_entity := $Rock as res://rock.tscn` (or something like that)
without the need to write `class_name Rock` in the rock scene.

**Reply by AThousandShips** [reply]
You can already do this, by doing:
```gdscript
const rock_type = preload("rock.gd")
...
$Rock as rock_type
```

See [here](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#variables)

**Reply by AThousandShips** [reply]
You also should be re-using the same standard code, if you are attatching unique scripts to each single rock you're probably doing something wrong

Note that a scene is not a type

**Reply by SKFrozenCloud** [reply]
I'm reusing code but the problem was the need to put `class_name` on every single custom node.

The solution works but it's two lines. It would be nice if it could be written as:
`var rock := $Rock as preload("rock.gd")`

Otherwise for each reference it needs double the amount of lines of code.

EDIT:

It works, but are there any performance implications of preload?

**Reply by AThousandShips** [reply]
Why is it a problem to use `class_name`? How many different custom nodes do you have that this is not usable? You're already naming the file so it's not a problem of coming up with a name, and it's a single line of code

I don't see why this is a problem?

> It would be nice if it could be written as: 
`var rock := $Rock as preload("rock.gd")`

Agreed

**Reply by AThousandShips** [reply]
> It works, but are there any performance implications of preload?

Likely yes, and it limits certain things like circular references, which is why `class_name` is just better

**Reply by SKFrozenCloud** [reply]
Yeah, you are right, the drawbacks outweight the benefits.

`class_name` seems to be the best solution.
